pandas/core/indexes/datetimelike.py

```diff
@@ -168,12 +169,14 @@ def __contains__(self, key: Any) -> bool:
             is_scalar(res) or isinstance(res, slice) or (is_list_like(res) and len(res))
         )
 
-    def sort_values(self, return_indexer=False, ascending=True):
+    def sort_values(self, return_indexer=False, ascending=True, key=None):
         """
         Return sorted copy of Index.
         """
+        idx = ensure_key_mapped(self, key)
```

Comment:
I noticed below that `ensure_key_mapped` is actually converting back to the original type. Which is good in general, but for index means it converts back to the original Index subclass, while I think we should probably use the base class `Index(..)` constructor in that case.

But I am also fine to leave this as a follow-up enhancement.

Reply:
I fixed this. See what you think about the current version. Basically, .asi8 ignores the type, so it doesn't matter. This works as intended, but it's a _little_ bit of a hack.

Reply:
Can you add a test for this?

`asi8` returns None for the base Index class, so I don't think it will work for all cases. For example, assume you would sort a DatetimeIndex on the name of the month (a bit contrieved, but to show the point): `key=lambda idx: idx.month_name()`)

Reply:
Right that makes sense. Could we just always use the first branch for `return_indexer` instead of the second branch that uses `_simple_new`? I'm not sure what the whole section with `freq` is trying to do. Performance optimization?

Reply:
As you mentioned above, this does work:

```python
    if not isinstance(result, type_of_values):  # recover from type error
        try:
            if isinstance(values, Index):
                result = Index(result)
            else:
                result = type_of_values(result)
        except TypeError:
            raise TypeError("User-provided `key` function returned an invalid type.")
```

The code is a little ugly, but it works for any kind of Index. Thoughts?